### ISSUE TITLE
[BUG] bottommost level garbage key versions leak forever after compaction

### ISSUE THREAD

**Issue by wolfkdy** [facebook/rocksdb#11188]
> Note: Please use Issues only for bug reports. For questions, discussions, feature requests, etc. post to dev group: https://groups.google.com/forum/#!forum/rocksdb or https://www.facebook.com/groups/rocksdb.dev

### Expected behavior
Timestamped keys in bottommost level pined by full_history_ts_low should have a chance to be compacted when full_history_ts_low  moves forward

### Actual behavior
Follow the steps below, you will find the keys in bottommost level pined by full_history_ts_low leak

### Steps to reproduce the behavior
Step1: set full_history_low to 0 to pin all history
Step2: put some keys with timestamp=1, and then delete them all with timestap=1
Step3: Flush the keys to l0 and CompactRange to ensure the keys got into bottommost level
Step4:  assert there are some files in bottommost files, the files all have largest_seq = 0, so the compacted output files will never got another chance to be garbage-collected.

Explainations to Step4:
as implemented below 
https://github.com/facebook/rocksdb/blob/27cf09172c4f212d0aef23b23c802bb791ec9209/db/version_set.cc#L3988-L4000
files in bottommost files can be compacted iff fd.largest_seqno > 0

### Suggestion to fix this bug
Add minTs and maxTs to file's meta data, each time full_history_low moves forward, re-caculate files which need compaction.

### example code
```
TEST_F(TimestampCompatibleCompactionTest, BottommostLevelCompactionTest) {
  Options options = CurrentOptions();
  options.env = env_;
  options.sst_partitioner_factory =
      std::make_shared<TestFilePartitionerFactory>();
  options.comparator = test::BytewiseComparatorWithU64TsWrapper();

  constexpr int kNumFiles = 4;
  options.level0_file_num_compaction_trigger = kNumFiles;

  DestroyAndReopen(options);

  uint64_t full_history_low = 0;
  ASSERT_OK(db_->IncreaseFullHistoryTsLow(db_->DefaultColumnFamily(), Timestamp(full_history_low)));

  constexpr int kKeysPerFile = 2;
  const std::string user_key = "foo";
  constexpr uint64_t start_ts = 1;

  uint64_t cur_ts = start_ts;
  // Generate some initial files in both L0 and L1.
  for (int k = 0; k < kNumFiles; ++k) {
    for (int i = 0; i < kKeysPerFile; ++i) {
      ASSERT_OK(db_->Put(WriteOptions(),
                         user_key + std::to_string(i),
                         Timestamp(cur_ts),
                         "v"));
      ++cur_ts;
    }   
    ASSERT_OK(db_->Flush(FlushOptions()));
  }

  for (int k = 0; k < kNumFiles; ++k) {
    for (int i = 0; i < kKeysPerFile; ++i) {
      ASSERT_OK(db_->Delete(WriteOptions(),
                         user_key + std::to_string(i),
                         Timestamp(cur_ts)));
      ++cur_ts;
    }   
    ASSERT_OK(db_->Flush(FlushOptions()));
  }

  ASSERT_OK(dbfull()->TEST_WaitForCompact());
  ASSERT_EQ(2 * kNumFiles * kKeysPerFile,
                NumTableFilesAtLevel(/*level=*/1, /*cf=*/0));
  
  std::vector<LiveFileMetaData> file_metadatas;
  db_->GetLiveFilesMetaData(&file_metadatas);
  for (auto const& file : file_metadatas) {
      ASSERT_EQ(0, file.largest_seqno);
  }
}
```

**Comment by cbi42** [comment]
Thanks for reporting. I don't think we rely on kBottommostFiles compaction as the main garbage collection method for user-defined timestamp (https://github.com/facebook/rocksdb/wiki/User-defined-Timestamp-%28Experimental%29#compaction-and-garbage-collection-gc). In the scenario you described, keys should still be garbage collected when an automatic compaction or a manual compaction in bottommost level overlaps with the key range.

**Comment by wolfkdy** [comment]
@cbi42 , thanks for your reply.

> keys should still be garbage collected when an automatic compaction or a manual compaction in bottommost level overlaps with the key range

As you said, 
1.  I agree that a manual compaction with BottommostLevelCompaction::kForce can successfully gc timesamped keys pined by full_history_ts_low, since manual compaction does not rely on `bottommost_files_marked_for_compaction_` when picking files.
2.  I agree that an auto compaction ***which overlaps with the garbage key range*** can succefully gc imesamped keys pined by full_history_ts_low, since they overlaped.
3.  However, if no futher writes overlap with ***bottommost level garbage timestamped keys*** , when and how will these keys be garbage collected? I think rocksdb should have a mechanism like `ComputeBottommostFilesMarkedForCompaction` , which garbage collects keys pinned by snapshots. For keys pinned by full_history_ts_low, I haven't find one.

Make a conclusion, for keys in the bottommost level which has largest_seq = 0 since they have been compacted once, but are still garbages because they were pinned by full_history_ts_low in the last compaction, and if no further writes overlap with them, I didn't find any mechanism that ***auto compaction*** can garbage collect them.
So, I still insist it's a bug

**Comment by cbi42** [comment]
>  I think rocksdb should have a mechanism like ComputeBottommostFilesMarkedForCompaction

I agree it would work as you suggested if we also record minTs and maxTs and check them in ComputeBottommostFilesMarkedForCompaction (actually they existed before but were removed as not being used https://github.com/facebook/rocksdb/pull/10443). I consider this more as a feature suggestion for garbage collection with user-defined timestamp though I don't think it's planned yet.

**Comment by ajkr** [comment]
Should the deletions have nonzero seqno? Having seqno zero (same as the keys they cover) while having larger timestamps seems to contradict the wiki's claim: "If ts1 < ts2, then seq1 < seq2"

**Comment by cbi42** [comment]
> Should the deletions have nonzero seqno? Having seqno zero (same as the keys they cover) while having larger timestamps seems to contradict the wiki's claim: "If ts1 < ts2, then seq1 < seq2"

I think currently deletions can have zero seqno, which violates the contract you quoted. I'm not sure whether this is intentional since read path seems still correct to me.
